### PR TITLE
Managers

### PR DESCRIPTION
**Pull request by MunWolf**
Refactored how we handle creating, updating, rendering and detect collisions on entities.

### CODE REVIEW

**Comment by MunWolf** [comment]
Also fixes #13

**Comment by MunWolf** [comment]
Hope im not making the code way too complicated for people here, not sure on the programming level of contribitors.

**Comment by MunWolf** [comment]
I am rebasing this and fixing it up btw.

File: entitymanager.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+  this.render = function() {
+    for (var index in entities) {
+      entities[index].render();
+    }
+  }
+
+  this.checkCollisions = function() {
+    var x = 1;
+    for (var i in entities) {
```
Comment (meiamsome) [inline entitymanager.js:27]: This would probably be better expressed with an old-fashioned loop:

    for(var i = 0; i < entities.length; i++) {
      for(var j = i + 1; j < entities.length; j++) {

Comment (MunWolf) [reply entitymanager.js:27]: This is not possible as entities is not an array but an object. I do this so every entity has a consistent id in the entitymanager, aka even if an entity dies the objects that were made after it don't change their ids.

Comment (meiamsome) [reply entitymanager.js:27]: Ah, okay, I obviously didn't read thoroughly enough!

**Comment by MunWolf** [comment]
Ok everything should be rebased and updated to play nice with everyone elses changes.

File: entitymanager.js (hunk 1 above)
Comment (MunWolf) [reply entitymanager.js:27]: No problem, if you can think of a better way to loop through this without getting duplicate pairs and not changing entities to an array do tell me :)

P.S. doing the for loop with the indices would work if you used the nextId variable but you would get alot of undefined properties and that gets worse and worse as we play the game longer.

P.S.S. The reason I am not using an array is that I am using the id in the register listener function and deregister so I can determine which listeners I should remove when an object is destroyed and if I used splice I could register two objects under the same id.

**Comment by shiffman** [comment]
So I love this and am amazed by it.  But I do wonder if it's too much for this repo as the game is meant to be a simple system for beginners to be able to follow and add features to.  I don't want this work to go to waste.  Do you think it might make sense to create a forked version that lives somewhere with a fancier manager and is liked from README.txt?   What do you think?  Thank you so much for your participation!

**Comment by MunWolf** [comment]
Honestly it kinda went out of hand, I just meant to fix #13 and then I suddenly started refactoring everything, it depends what you think is harder for new programmers to understand I guess.

I can easily just have this as a fork if you think it is better to have there :)

P.S. If you want me to do the fork thing just close this PR and I will make a followup where I change the README with a link to mine.

**Comment by LeastCreative** [comment]
I have been programming for years but I am a beginner to javascript (minus everyday jquery stuff we use .NET at work) and I have already learned a ton just from the other people here.

I think this project should get as awesome as possible! The original version is still out there for people following the videos

**Comment by ada-lovecraft** [comment]
I've been weighing the **Complexity vs. Correctness** of this project. I've been tempted to sit down and add object pooling, hud managers, abstract sprite classes, and a whole host of other things.

The problem I see, is that, with enough time and engineering, we're not gonna build a **game** but an entire game *framework*, which, as I see it, isn't the point of this. 

That being said, I like where @WolfGaming is going, and am actually working on addressing some of these issues in my next (read: *first*) video podcast as a reaction to some of the things @shiffman did in his live show this week.

The bottom line is: this project is not *for* those of us who are advanced coders. It is intended for beginners and experimenters. I'd rather not scare aware newbies by over-architecting this repo. 

## The Proposal
Let's create a duplicate repository called **Asteroids++** and make it a place where the advanced developers can play around in. This way, we can satisfy the conceit of the original project, and have a place where those of us who wish to, can -really- let loose.

I've got everything ready to go. Just say when

**Comment by kwichmann** [comment]
Sounds like a great idea @codevinsky. Also, as you do all of this stuff the correct way, a few basic tutorials along the way for the newbs like yours truly would be super cool. Maybe in some of your upcoming vids (which I'm looking forward to, btw)?

**Comment by shiffman** [comment]
@codevinsky sounds great!  You are welcome to make a repo here or in your own account, whatever you prefer!

**Comment by shiffman** [comment]
Can this be closed now that it's a feature of "advanced" asteroids?

**Comment by MunWolf** [comment]
Yes, closing it now.
Here is the new request for anyone interested
https://github.com/CodingRainbow/asteroids-advanced/pull/3
